torchdynamo/optimizations/training.py

```diff
@@ -25,6 +25,60 @@
 log = logging.getLogger(__name__)
 
 
+def is_aot_autograd_safe_to_run(gm, example_inputs):
+    """
+    There are some known issues with Aot Autograd. This is a workaround to catch
+    such cases, and fallback to eager. We should fix these quickly.
+
+    Issues
+    1) LSTM - https://github.com/pytorch/torchdynamo/issues/1147
+    2) LSTM - https://github.com/pytorch/functorch/issues/586
+    3) TODO - Add an issue for set_grad_enabled
+    4) Input mutation - https://github.com/pytorch/torchdynamo/issues/1301
+    """
+
+    def raise_and_false():
+        if config.raise_on_unsafe_aot_autograd:
+            raise NotImplementedError("Aot Autograd is unsafe")
+        return False
+
+    import functorch.compile
+
+    # 1) LSTM module (tts_angular) - https://github.com/pytorch/functorch/issues/586
+    for submod in gm.modules():
+        if submod.__class__.__name__ == "LSTM":
+            log.warning("Unable to use Aot Autograd because of presence of LSTM")
+            return raise_and_false()
+
+    # 2) set_grad_enabled
+    for node in gm.graph.nodes:
+        if node.target == torch._C._set_grad_enabled:
+            log.warning(
+                "Unable to use Aot Autograd because of presence of set_grad_enabled call"
+            )
+            return raise_and_false()
```

Comment:
nit: you might as well have `raise_and_false` take a `reason` and then move the log line in:

`log.warning(f"Unable to use Aot Autograd because of {reason}")`
...
`return raise_and_flag("presence of LSTM")`
`return raise_and_flag("set_grad_enabled call")`
etc

Reply:
Nice!